faststreams/input_stream.nim

```diff
@@ -1,192 +1,298 @@
-
-proc endPos*(s: InputStreamVar): int =
-  # TODO This needs to use a VTable for `system.File` based streams
+    stream.bufferStart = head
+
+  InputStreamHandle(s: stream)
+
+proc memoryInput*(mem: openarray[byte]): InputStreamHandle =
+  let head = unsafeAddr mem[0]
+  InputStreamHandle(s: InputStream(
+    head: head,
+    bufferEnd: offset(head, mem.len),
+    bufferEndPos: mem.len))
+
+proc memoryInput*(str: string): InputStreamHandle =
+  memoryInput str.toOpenArrayByte(0, str.len - 1)
+
+# TODO: Is this used, should we deprecate it?
+proc endPos*(s: InputStream): int =
+  doAssert s.vtable == nil or s.vtable.getLenSync != nil
   return s.bufferEndPos
 
-proc syncRead(s: var ByteStream): bool =
-  let bytesRead = s.reader(s.bufferStart, s.bufferSize)
+# TODO The return type here could be Option[Natural] if Nim had
+# the Option[range] optimisation that will make it equvalent to `int`.
+proc len*(s: InputStream): int {.raises: [Defect, IOError].} =
+  if s.vtable == nil:
+    distance(s.head, s.bufferEnd)
+  elif s.vtable.getLenSync != nil:
+    s.vtable.getLenSync(s)
+  else:
+    lengthUnknown
+
+template len*(s: AsyncInputStream): int =
+  len InputStream(s)
+
+proc bufferMoreDataSync(s: InputStream): bool =
+  # Returns true if more data was successfully buffered
+  if s.vtable == nil or s.vtable.readSync == nil:
+    return false
+
+  let bytesRead = s.vtable.readSync(s, s.bufferStart, s.bufferSize)
   if bytesRead == 0:
-    s.reader = nil
-    return true
+    # TODO close the input device
+    s.vtable = nil
+    return false
   else:
     s.bufferEnd = offset(s.bufferStart, bytesRead)
     s.bufferEndPos += bytesRead
-    return false
+    return true
 
-proc ensureBytes*(s: var ByteStream, n: int): bool =
+proc bufferMoreDataAsync(s: AsyncInputStream): Future[bool] {.async.} =
+  # Returns true if more data was successfully buffered
+  return false
+
+proc readable*(s: InputStream): bool =
+  if s.head != s.bufferEnd:
+    true
+  else:
+    s.bufferMoreDataSync()
+
+template readable*(sp: AsyncInputStream): bool =
+  let s = sp
+  if s.head != s.bufferEnd:
+    true
+  else:
+    faststreamsAwait bufferMoreDataAsync(s)
+
+proc readable*(s: InputStream, n: int): bool =
   if distance(s.head, s.bufferEnd) >= n:
     return true
 
-  if s.reader == nil:
+  if s.vtable == nil or s.vtable.readSync == nil:
     return false
 
+  # TODO
   doAssert false, "Multi-buffer reading will be implemented later"
 
-proc eof*(s: var ByteStream): bool =
-  if s.head != s.bufferEnd:
-    return false
+template readable*(sp: AsyncInputStream, n: int): bool =
+  let s = sp
 
-  if s.reader == nil:
+  if distance(s.head, s.bufferEnd) >= n:
     return true
 
-  return s.syncRead()
+  if s.vtable == nil:
+    return false
+
+  # TODO
+  doAssert false, "Multi-buffer reading will be implemented later"
 
-proc eob*(s: ByteStream): bool {.inline.} =
-  s.head != s.bufferEnd
+template close*(s: AsyncInputStream) =
```

Comment:
closing an async stream is async

Reply:
I haven't quite finished this part. One thing to note is that no language has invented async destructors yet, so I'm still thinking what would be the best ways to enforce closing in the async scenario.

Reply:
well, async close/destructor is a bit of a funny feature, in general - my opinion would be that `close` shouldn't be async, but unfortunately it is, in chronos.

As an example bsd sockets have `shutdown` to implement two-phase async closes but nobody uses it in practise - by necessity you already have to handle the case where you "don't know" if the thing was closed successfully  or not, and that logic is exactly the same as cleaning up after a failed destruction (usually done on the next connection/open), so at that point, it's very rare that you want to bother to "properly" close things and act on the outcome - thus it makes sense that you have a separate API to performed a "controlled" shutdown.

If you add "controlled shutdown", your stream should also have a full lifecycle around "closing" - it then becomes "just another" state that the stream may be in (ie closed, opening, open, closing, closed, repeat).

Mostly, I see close as an "advisory" feature: a tiny optimization to release some resources a little bit earlier, which is about the level of quality that generally is useful. This is a bit of a problem in stream interfaces in general: if you want to cover both sockets and files, the expectations around "close" are different and you have to choose the lowest common denominator (non-deterministic close) or more-complexity-for-all (async close, even though for many cases it "could" be sync, like files most of the time)

Reply:
My own thinking so far is that the user should be able call `close` manually and it offers both sync and async versions. If you call it manually, you can handle the errors. If you don't call it manually, then it's executed in sync form in the destructor. Any errors that might happen in the destructor are optionally passed along to a global hook installed by the application. If no such hook was installed, they are ignored because you cannot raise exceptions in destructors.

Reply:
global hooks (like the global hook for exceptions that nobody uses) are a mess to implement correctly and efficiently in the compiler, fwiw - they're fundamentally crippled like any other mutable global specially if you want to support libraries and not just full-source-code views of the world.. it would be better if it lived in the type like `unique_ptr`

Reply:
This misses the point. If you don't close the stream explicitly and rely on a destructor-like mechanism, how do you handle the errors in the destructor? You cannot throw an exception there and logging is not appropriate in a library like FastStreams. You can decide to ignore the error by default, but the hook at least gives the user a chance to install their own logging in there. I'm open to suggestions for other ideas that solve this problem.